File: exercise/145/morris.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void reverse(vector<int> &result, int left, int right) {
        for (int i = 0; i < (right - left) / 2; i++) {
            int t = result[left + i];
            result[left+i] = result[right-i-1];
            result[right-i-1] = t;
        }
    }
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> result;
        TreeNode fake(0);
        fake.left = root;
        TreeNode *p = &fake;
        while (p) {
            if (!p->left) {
                p = p->right;
                continue;
            }
            TreeNode *pre = p->left;
            while (pre->right && pre->right != p) {
                pre = pre->right;
            }
            if (pre->right == p) {
                pre->right = NULL;
                int a = result.size();
                pre = p->left;
                while (pre) {
                    result.push_back(pre->val);
                    pre = pre->right;
                }
                reverse(result, a, result.size());
                p = p->right;
            } else {
                pre->right = p;
                p = p->left;
            }
        }
        return result;
    }
};

File: exercise/145/morris.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def reverseResult(self, result, begin, end):
        for i in range((end - begin) // 2):
            result[begin+i], result[end-1-i] = result[end-i-1], result[begin+i]

    def postorderTraversal(self, root: TreeNode) -> List[int]:
        result = []
        p = root
        while p:
            if p.left is None:
                p = p.right
            else:
                pre = p.left
                while pre.right and pre.right != p:
                    pre = pre.right
                if pre.right == p:
                    pre.right = None
                    now = len(result)
                    output_p = p.left
                    while output_p != pre:
                        result.append(output_p.val)
                        output_p = output_p.right
                    result.append(pre.val)
                    self.reverseResult(result, now, len(result))
                p = p.right
                else:
                    pre.right = p
                    p = p.left
        return result

